sdk/nodejs/resource.ts

```diff
@@ -335,21 +353,31 @@ export interface Alias {
 
 // collapseAliasToUrn turns an Alias into a URN given a set of default data
 function collapseAliasToUrn(
-    alias: Input<Alias | string>, defaultName: string, defaultType: string,
-    defaultParent: Resource | undefined): Output<URN> {
+        alias: Input<Alias | string>,
+        defaultName: string,
+        defaultType: string,
+        defaultParent: Resource | undefined): Output<URN> {
 
     return output(alias).apply(a => {
         if (typeof a === "string") {
             return output(a);
-        } else {
-            return createUrn(
-                output(a.name).apply(n => n || defaultName),
-                output(a.type).apply(ty => ty || defaultType),
-                a.parent || defaultParent,
-                a.project,
-                a.stack,
-            );
         }
+
+        const name = a.hasOwnProperty("name") ? a.name : defaultName;
+        const type = a.hasOwnProperty("type") ? a.type : defaultType;
+        const parent = a.hasOwnProperty("parent") ? a.parent : defaultParent;
+        const project = a.hasOwnProperty("project") ? a.project : getProject();
+        const stack = a.hasOwnProperty("stack") ? a.stack : getStack();
```

Comment:
Is there any benefit to changing any of these other than `parent`?  I would have assumed we want to just make a targeted change to how `parent` is interpreted since it is the only one where there is are two meaningful defaults we want to allow the user to distinguish between.

Reply:
i found it much easier to reason about if they all behave the same way.

Reply:
Do we need to update the documentation here then, to call out that `name: undefined` is almost certainly wrong?

Actually, for all of these expect `parent` what are the cases where the new behavior is "right"?  It seems like if you actually have a key in the property bag but it is set to undefined very wacky things are going to happen.

Reply:
I'd prefer to not think about it as 'right' or 'wrong'.  Rather, i'd like to think about it as "consistent" or "inconsistent".  Right now, there's no need to describe different behavior.  It's simply all the same.  If you provide properties, we use them.  If you don't we use the existing value for the Resource.  I agree that only `parent: undefined` makes sense for actually passing `undefined`, but i'm ok with that.  If you happen to pass `name: undefined` it will throw as something invalid.